### PR TITLE
Monitor Query

### PR DESCRIPTION
**Pull request by gracewilcox**
Fixes https://github.com/Azure/azure-sdk-for-go/issues/18776

### CODE REVIEW

File: sdk/monitor/azquery/example_test.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	_ = client
+}
+
+func ExampleLogsClient_QueryWorkspace() {
+	client := azquery.NewLogsClient(cred, nil)
+	timespan := azquery.TimeInterval(time.Now().Add(-12*time.Hour), time.Now())
+
+	res, err := client.QueryWorkspace(context.TODO(), workspaceID,
+		azquery.Body{Query: to.Ptr(query), Timespan: to.Ptr(timespan)}, nil)
+	if err != nil {
```
Comment (richardpark-msft) [inline sdk/monitor/azquery/example_test.go:51]: I think the general pattern is to just panic now. So you can just add in:

```go
if err != nil {
   panic(err)
}
```

File: sdk/monitor/azquery/example_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		// TODO: handle error
+	}
+	_ = res
```
Comment (richardpark-msft) [inline sdk/monitor/azquery/example_test.go:54]: It can be nice to show some usage here, especially for common fields.

Comment (gracewilcox) [reply sdk/monitor/azquery/example_test.go:51]: plan on adding to documentation

File: sdk/monitor/azquery/example_test.go (hunk 1 above)
Comment (gracewilcox) [reply sdk/monitor/azquery/example_test.go:48]: updated to panic

File: sdk/monitor/azquery/custom_client.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+type MetricsClientOptions struct {
+	azcore.ClientOptions
+}
+
+// LogsClientOptions contains optional settings for LogsClient.
+type LogsClientOptions struct {
+	azcore.ClientOptions
+}
+
+// NewLogsClient creates a client that accesses a monitor.
```
Comment (richardpark-msft) [inline sdk/monitor/azquery/custom_client.go:27]: Is 'monitor' the name of the resource? Is it capitalized or is there a more official name for the resource type?

File: sdk/monitor/azquery/metrics_client.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+// NewListMetricDefinitionsPager - Lists the metric definitions for the resource.
+// Generated from API version 2018-01-01
+// resourceURI - The identifier of the resource.
+// options - MetricsClientListMetricDefinitionsOptions contains the optional parameters for the MetricsClient.ListMetricDefinitions
+// method.
+func (client *MetricsClient) NewListMetricDefinitionsPager(resourceURI string, options *MetricsClientListMetricDefinitionsOptions) *runtime.Pager[MetricsClientListMetricDefinitionsResponse] {
+	return runtime.NewPager(runtime.PagingHandler[MetricsClientListMetricDefinitionsResponse]{
+		More: func(page MetricsClientListMetricDefinitionsResponse) bool {
```
Comment (richardpark-msft) [inline sdk/monitor/azquery/metrics_client.go:34]: How much desire do we have to rename or shorten some of these? Just the fact that it has `Metric` in the name is probably enough and we could omit the "MetricsClient" part of the type altogether.

**Review by richardpark-msft — APPROVED** [review]
I don't see anything too bad or objectionable. 

For the next beta I think we should concentrate on getting those examples so they demonstrate some bits of Kusto, and display/print out some of the results. Any time there's an any/interface{} type you need to show how to pick it apart.

Take a look at the comments, fix any you think you can do reasonably and just file issues for the rest for beta.2.

File: sdk/monitor/azquery/README.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+### Authentication
+
+This document demonstrates using [azidentity.NewDefaultAzureCredential][default_cred_ref] to authenticate. This credential type works in both local development and production environments. We recommend using a [managed identity][managed_identity] in production.
+
+[Client][client_docs] accepts any [azidentity][azure_identity] credential. See the [azidentity][azure_identity] documentation for more information about other credential types.
+
+#### Create a logs client
+
+```go
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:36]: Noted, will migrate to example_test file next beta

File: sdk/monitor/azquery/README.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/monitor/azquery"
+)
+
+func main() {
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		// TODO: handle error
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:45]: Will be fixed when I migrate to example files. Already fixed there

File: sdk/monitor/azquery/README.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+```go
+client := azquery.NewLogsClient(cred, nil)
+timespan := "2022-08-30/2022-08-31"
+additionalWorkspaces := []*string{&workspaceID2, &workspaceID3}
+
+res, err := client.QueryWorkspace(context.TODO(), workspaceID, azquery.Body{Query: to.Ptr(query), Timespan: to.Ptr(timespan), Workspaces: additionalWorkspaces}, nil)
+if err != nil {
+	// TODO: handle error
+}
+_ = res
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:234]: will add when moved to example

File: sdk/monitor/azquery/README.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+```
+
+#### Batch query request structure
+
+```
+BatchRequest
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:185]: Just wanted the users to have an easy reference for what's required for the Batch Request as it differs a lot from the single query request

File: sdk/monitor/azquery/README.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+	|---ErrorMessage *string
+|---Cost *int32
+|---Interval *string
+|---Namespace *string
+|---Resourceregion *string
+```
+
+## Troubleshooting
+
+More troubleshooting steps coming soon.
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:311]: removing now

File: sdk/monitor/azquery/README.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+<!-- LINKS -->
+[managed_identity]: https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview
+[azure_identity]: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity
+[azure_sub]: https://azure.microsoft.com/free/
+[azure_monitor_create_using_portal]: https://docs.microsoft.com/azure/azure-monitor/logs/quick-create-workspace
+[azure_monitor_overview]: https://docs.microsoft.com/azure/azure-monitor/overview
+[azure_subscription]: https://azure.microsoft.com/free/java
+[changelog]: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/monitor/azure-monitor-query/CHANGELOG.md
+[DefaultAzureCredential]: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity/README.md#defaultazurecredential
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:353]: fixed- removed

File: sdk/monitor/azquery/README.md [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+	// TODO: handle error
+}
+_ = res
+```
+
+#### Increase wait time, include statistics, include render (visualization)
+
+By default, the azure service will run your query for up to three minutes. To increase the default timeout, update the wait time in seconds in LogsClientQueryWorkspaceOptions Prefer string. Max wait time the service will allow is ten minutes.
+
+To get logs query execution statistics, such as CPU and memory consumption, set include-statistics to true in LogsClientQueryWorkspaceOptions Prefer string.
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:241]: fixed

File: sdk/monitor/azquery/README.md [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+res, err := client.QueryWorkspace(context.TODO(), workspaceID, azquery.Body{Query: to.Ptr(query), Timespan: to.Ptr(timespan), Workspaces: additionalWorkspaces}, nil)
+if err != nil {
+	// TODO: handle error
+}
+_ = res
+```
+
+#### Increase wait time, include statistics, include render (visualization)
+
+By default, the azure service will run your query for up to three minutes. To increase the default timeout, update the wait time in seconds in LogsClientQueryWorkspaceOptions Prefer string. Max wait time the service will allow is ten minutes.
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:239]: fixed

File: sdk/monitor/azquery/custom_client.go (hunk 3 above)
Comment (gracewilcox) [reply sdk/monitor/azquery/custom_client.go:27]: fixed

File: sdk/monitor/azquery/metrics_client.go (hunk 4 above)
Comment (gracewilcox) [reply sdk/monitor/azquery/metrics_client.go:34]: Good point, added to todo list for next release

File: sdk/monitor/azquery/README.md [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+### Metrics data structure
+
+Each set of metric values is a time series with the following characteristics:
+
+- The time the value was collected
+- The resource associated with the value
+- A namespace that acts like a category for the metric
+- A metric name
+- The value itself
+- Some metrics may have multiple dimensions as described in multi-dimensional metrics. Custom metrics can have up to 10 dimensions.
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:91]: Good point, on todo list for next release

File: sdk/monitor/azquery/README.md [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+	// TODO: handle error
+}
+_ = res
+```
+
+### Metrics query
+
+```go
+client := azquery.NewMetricsClient(cred, nil)
+res, err := client.QueryResource(context.Background(), resourceURI,
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:263]: todo for next release

File: sdk/monitor/azquery/README.md [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+	|---Body *BatchQueryResults
+		|---Error *ErrorInfo
+		|---Render interface{}
+		|---Statistics interface{}
+		|---Tables []*Table
+			|---Columns []*Column
+				|---Name *string
+				|---Type *LogsColumnType
+			|---Name *string
+			|---Rows [][]interface{}
```
[the remark this answers is not in the document]
Comment (gracewilcox) [reply sdk/monitor/azquery/README.md:211]: Making rows clearer is a todo for next release. Rows list out the results of the query

File: sdk/monitor/azquery/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (scottaddie) [reply sdk/monitor/azquery/README.md:36]: That's correct. This work is slated for Zinc. It's something Rick, Charles, and I discussed.

File: sdk/monitor/azquery/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (scottaddie) [reply sdk/monitor/azquery/README.md]: Once you have created a troubleshooting guide, it would make sense to add this section back and link to it.

**Pull request merged**